### ISSUE TITLE
torch.distributed.run to discover an available port automatically

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#73320]
### 🚀 The feature, motivation and pitch

Currently two lauches on the same machine (e.g. one run uses devices 0-3, and another run uses 4-7) often result in "Address already in use". On Linux torch should be able to find an available ephemeral port automatically. Eph

**Comment by zhaojuanmao** [comment]
cc @kiukchung

**Comment by zhaojuanmao** [comment]
cc @cbalioglu, does TCPStore automatically discover an available port?

**Comment by cbalioglu** [comment]
@vadimkantorov If you are running more than one distributed job, you have to have some form of information that uniquely identifies each job. Otherwise technically it is not possible for workers to figure out which job they belong to. Starting with v1.9 `python -m torch.distributed.run` internally uses TorchElastic which uses a fixed port number (defaults to 29500) for the rendezvous among worker processes. This means implicitly that port number "identifies" the job instance. If you want to have secondary jobs running on the same node(s), you have to manually specify a port number:

```
python -m torch.distributed.run --master_port <port_number>
```

**Comment by vadimkantorov** [comment]
Why can't the port be randomly selected by the launcher? E.g. it may perform similarly `--master_port $RANDOM # do it properly from the ephemeral local port range` This is already happening for back-connects. At least providing such an option would be very useful "--master_port_random_available"

**Comment by cbalioglu** [comment]
I agree that the setup can be slightly simplified for single node jobs. @aivanou or @kiukchung (owners of TorchElastic) might chime in here to address your feedback.

**Comment by kiukchung** [comment]
@vadimkantorov, @cbalioglu for pure local runs it makes sense to have torch.distributed.run select a random port for convenience. We can make it in such a way that —master_port=0 is passed down to TCPStore, which I believe we already made it listen on an available random port of port=0.

**Comment by kiukchung** [comment]
I just verified that passing `--master_port=0` with `--rdzv_backend=c10d` will assign a random port already. This would only work for single-node, multi-worker usecases. And you'd also have to manually mask GPUs by setting `CUDA_VISIBLE_DEVICES` accordingly (if you are using `env["LOCAL_RANK"]` to set the device in your training script).

```
$  torchrun --nproc_per_node 2 --nnodes 1 --rdzv_backend c10d --master_port 0 test.py 
```

**Comment by vadimkantorov** [comment]
@kiukchung I think that this mode should be more explicit or even default. This is the most frequent way of using the distributed regime. The workaround you proposed is quite complex and easy to get wrong. I think that this mode should be the one described first in all tutorials, since it's the least demanding of machines / resources (especially for academic environments).

**Comment by kiukchung** [comment]
thanks for the feedback, I'll add a section in the docs to help users discover `--master_port=0` for multiple simultaneous runs in local mode. As for making `master_port` by default 0, is not a decision we can take lightly since it would technically break backwards compatibility and we have seen many usages of `torchrun` in a multi-node context.

Slightly off-topic (but I find it relevant enough to mention here). This is why we started a project called TorchX (see the relevant ddp section here: https://pytorch.org/torchx/latest/components/distributed.html). In TorchX we can freely create different "flavors" of `torchrun` that hard codes (templetizes) these configuration groups for different use-cases. For instance we could run a multi-node distributed `torchrun` as:

```
$ torchx run dist.ddp --script trainer.py -j 4x8 # 4 nodes 8 workers
```

and a local version (that allows for local stacking) as:

```
# 1st copy
$ torchx run local.ddp --script trainer.py -j 1x4 
# 2nd copy
$ torchx run local.ddp --script trainer.py -j 1x4 
```

In all cases this would invoke `torch.distributed.run` but with different templetizations of configurations that can be referenced by a human readable "name" (e.g `dist.ddp` versus `local.ddp`). For the local mode, we would also take care of setting `CUDA_VISIBLE_DEVICES` and such.

**Comment by vadimkantorov** [comment]
Well, if `--master_port=0` is the single place, it'd be a great and intuitive way for single-node.

The others: specifying rdv method and manually settting gpu device are much less intuitive and easy to mess up.

**Comment by kiukchung** [comment]
oops one correction, the option should've been `--rdzv_endpoint localhost:0` not `--master_port 0`, so 

```
$  torchrun --nproc_per_node 2 --nnodes 1 --rdzv_backend c10d --rdzv_endpoint localhost:0 test.py 
```

**Comment by vadimkantorov** [comment]
I would argue that this most useful, most basic and most frequent single-node multi-gpu regime deserves a  simple support into torch.distributed.run directly (and hopefully not even requiring the magic wrapper torchrun)

**Comment by kiukchung** [comment]
torchrun isn't a magic wrapper, its just a python console_entrypoint (see: https://github.com/pytorch/pytorch/blob/master/setup.py#L861) added for convenience (`torchrun` versus `python -m torch.distributed.run`)

**Comment by vadimkantorov** [comment]
Even if it's not magic, the other points still are relevant IMHO. If `--master_port 0` is sufficient to enable random port selection, it would be perfect!

**Comment by kiukchung** [comment]
I see, yes, thanks for pointing out the lack of documentation regarding the difference between `torchrun` and `torch.distributed.run`. I've clarified this in https://github.com/pytorch/pytorch/pull/73598 as well as added a section for stacked, single-node, multi-worker usecase.

**Comment by vadimkantorov** [comment]
@kiukchung As you see, people it's not clear to users that `torchrun` is a synonym for `python -m torch.distributed.run`: https://github.com/pytorch/pytorch/issues/73656 and that it's torch.distributed.run that should be checked for modifications.

Maybe would be good for torch.distributed.run to print out it's own arguments or add a print indicating that it's torch.distributed.run that's running.

**Comment by vadimkantorov** [comment]
@kiukchung currently it's required to set at least three parameters `--rdzv_backend=c10d --rdzv_endpoint=localhost:0 --nnodes=1`. Is there any tradeoff of always having `--rdzv_backend=c10d --rdzv_endpoint=localhost:0` by default? then if set `--nnodes=1`, it very likely means that a random port is wanted

**Comment by adam-dziedzic** [comment]
> --master_port 25901

Just for future such issues: I was running PyTorch distributed training on a machine with another PyTorch process (that I wasn't able to kill or stop). I was able to run my job by changing the port from default 25900 to 25901.

**Comment by HKEa** [comment]
if you could add more explicit instruction for the random port auto selection. It's not clear enough from the pytorch site https://pytorch.org/docs/stable/elastic/run.html
thanks for the great work

**Comment by vadimkantorov** [comment]
There is also some related workarounds in detectron2: https://github.com/facebookresearch/detectron2/blob/main/detectron2/engine/launch.py#L15?

If these are not needed, detectron2 could remove this code

**Comment by vadimkantorov** [comment]
@malfet could we reopen this issue? to repurpose for torchrun to discover an available port automatically when `--standalone` is used? (for the case of running multiple small runs on a single node: e.g. one person uses GPUs 0-3, another person GPUs 4-7)

So that tricky options like `--rdzv_backend c10d --rdzv_endpoint localhost:0 ` need not be used (and btw they are not mentioned at https://docs.pytorch.org/docs/stable/elastic/run.html#stacked-single-node-multi-worker)
